Items found on day 1 of advanced course (#333)
hsf-training/cpluspluscourse: talk/concurrency/threadlocal.tex

\subsection[TLS]{Thread-local storage}

\begin{frame}[fragile]
  \frametitlecpp[11]{Thread-local storage}
  \begin{block}{\mintinline{cpp}{thread_local} keyword}
    \begin{itemize}

>       \item When declaring a variable thread-local, every thread will have its own copy

Reviewer: ```suggestion
      \item A variable can be declared \mintinline{cpp}{thread-local}
      \item Then every thread will have its own copy
```

>       \item Most useful for things treated as ``working memory'' in each thread

Reviewer: ```suggestion
      \item Most useful for ``working memory'' in each thread
```

>       \item Note: Can also be declared directly on the stack of a thread (faster)

Reviewer: ```suggestion
      \item Note: Can also be declared directly on the stack of a thread
      \begin{itemize}
        \item and will be faster, thus should be preferred when possible
      \end{itemize}
```

Reviewer 2: you might also want to mention that `thread_local` variables are (usually) implicitly `static` -- if you think about it it must be that way, but if you don't think about it you'll get surprising behavior :)

Reviewer 2: `thread_local` implies `static` for variables at block scope (which in my experience is the situation in which you most commonly use `thread_local`)

    \end{itemize}
  \end{block}
  \begin{exampleblock}{}
    \begin{cppcode*}{}
      thread_local int a{0};

>       std::thread t1([&](){ a++; });
>       std::thread t2([&](){ a++; });

Reviewer 2: might also want to use a `jthread` instead 
```suggestion
      std::thread t1([&] { a++; });
      std::thread t2([&] { a++; });
```

Author: You are generally right, but the example is supposed to remind students of other examples which we are showing. There, we don't use `jthread` yet.
On second thought, though, I'll accept the suggestion, since this gives us the opportunity to show it in action.
